doc/developers/cython.rst

```diff
@@ -0,0 +1,143 @@
+
+  .. code-block:: python
+
+         # For GCC and for clang
+         %%cython --compile-args=-fopenmp --link-args=-fopenmp
+         # For Microsoft's compilers
+         %%cython --compile-args=/openmp --link-args=/openmp
+
+* To debug C code (e.g. a segfault), do use ``gdb`` with:
+
+  .. code-block:: bash
+
+         gdb --ex r --args python ./entrypoint_to_bug_reproducer.py
+
+* To have access to some value in place to debug in ``cdef (nogil)`` context, use:
+
+  .. code-block:: cython
+
+         with gil:
+             print(state_to_print)
+
+* Note that Cython cannot parse f-strings with ``{var=}`` expressions, e.g.
+
+  .. code-block:: bash
+
+         print(f"{test_val=}")
+
+* scikit-learn codebase has a lot of non-unified (fused) types (re)definitions.
+  There currently is `ongoing work to simplify and unify that across the codebase
+  <https://github.com/scikit-learn/scikit-learn/issues/25572>`_.
+  For now, make sure you understand which concrete types are used ultimately.
+
+* You might find this alias to compile individual Cython extension handy:
+
+    .. code-block::
+
+         # You might want to add this alias to your shell script config.
+         alias cythonX="cython -X language_level=3 -X boundscheck=False -X wraparound=False -X initializedcheck=False -X nonecheck=False -X cdivision=True"
+
+         # This generates `source.c` as as if you had recompiled scikit-learn entirely.
+         cythonX --annotate source.pyx
+
+* Using the ``--annotate`` option with this flag allows generating a HTML report of code annotation.
+  This report indicates interactions with the CPython interpreter on a line-by-line basis.
+  Interactions with the CPython interpreter must be avoided as much as possible in
+  the computationally intensive sections of the algorithms.
+  For more information, please refer to `this section of Cython's tutorial <https://cython.readthedocs.io/en/latest/src/tutorial/cython_tutorial.html#primes>`_
+
+    .. code-block::
+
+         # This generates a HTML report (`source.html`) for `source.c`.
+         cythonX --annotate source.pyx
+
+Tips for performance
+^^^^^^^^^^^^^^^^^^^^
+
+* Understand the GIL in context for CPython (which problems it solves, what are its limitations)
+  and get a good understanding of when Cython will be mapped to C code free of interactions with
+  CPython, when it will not, and when it cannot (e.g. presence of interactions with Python
+  objects, which include functions). In this regard, `PEP073 <https://peps.python.org/pep-0703/>`_
+  provides a good overview and context and pathways for removal.
+
+* Make sure you have deactivated `checks <https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/_build_utils/__init__.py#L68-L87>`_.
+
+* Always prefer memoryviews instead over ``cnp.ndarray`` when possible: memoryviews are lightweight.
+
+* Avoid memoryview slicing: memoryview slicing might be costly or misleading in some cases and
```

Comment:
FWIW I have an PR that should dramatically improve this. It almost certainly won't (and shouldn't!) make it into Cython 3, but hopefully won't be long after that.  Essentially, a lot of the reference counting turns out to be pretty pointless since you're usually incrementing and decrementing exactly the same object so it can be skipped completely.

So hopefully this advice can change sometime in the next year. But not yet

Reply:
Thanks for mentioning that, @da-woods!

Reply:
Thank you for mentioning this, @da-woods.